web/hooks/pull_request.go

```diff
@@ -0,0 +1,198 @@
+		ExternalRepositoryID: uint64(payload.GetRepo().GetID()),
+		Number:               uint64(payload.GetPullRequest().GetNumber()),
+	})
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			wh.logger.Debugf("Ignoring pull request review event for non-managed pull request #%d, in %s",
+				payload.GetPullRequest().GetNumber(), payload.GetRepo().GetFullName())
+		} else {
+			wh.logger.Errorf("Failed to get pull request from database %v", err)
+		}
+		return
+	}
+
+	course, err := wh.db.GetCourseByOrganizationID(uint64(payload.GetOrganization().GetID()))
+	if err != nil {
+		wh.logger.Errorf("Failed to get course from database: %v", err)
+		return
+	}
+	user, err := wh.db.GetUserByRemoteIdentity(&pb.RemoteIdentity{
+		RemoteID: uint64(payload.GetSender().GetID()),
+		Provider: "github",
+	})
+	if err != nil {
+		wh.logger.Errorf("Failed to get user from database: %v", err)
+		return
+	}
+	reviewer, err := wh.db.GetEnrollmentByCourseAndUser(course.GetID(), user.GetID())
+	if err != nil {
+		wh.logger.Errorf("Failed to get reviewer from database: %v", err)
+		return
+	}
+
+	// Only if the review is from a course teacher, do we set the pull request to approved.
+	if reviewer.IsTeacher() {
+		pullRequest.SetApproved()
+		wh.db.UpdatePullRequest(pullRequest)
+		wh.logger.Debugf("Pull request successfully approved for repository: %s", payload.GetRepo().GetFullName())
+	}
+}
+
+func (wh GitHubWebHook) handlePullRequestOpened(payload *github.PullRequestEvent) {
+	wh.logger.Debugf("Received pull request opened event for repository: %s, in organization: %s",
+		payload.GetRepo().GetName(), payload.GetOrganization().GetLogin())
+
+	repos, err := wh.db.GetRepositoriesWithIssues(&pb.Repository{RepositoryID: uint64(payload.GetRepo().GetID())})
+	if err != nil {
+		wh.logger.Errorf("Failed to get repository by remote ID %d from database: %v", payload.GetRepo().GetID(), err)
+		return
+	}
+	if len(repos) != 1 {
+		wh.logger.Debugf("Ignoring pull request opened event for unknown repository: %s", payload.GetRepo().GetFullName())
+		return
+	}
+	repo := repos[0]
+	if !repo.IsGroupRepo() {
+		wh.logger.Debugf("Ignoring pull request opened event for non-group repository: %s", payload.GetRepo().GetFullName())
+		return
+	}
+	wh.createPullRequest(payload, repo)
+}
+
+func (wh GitHubWebHook) handlePullRequestClosed(payload *github.PullRequestEvent) {
+	wh.logger.Debugf("Received pull request closed event for repository: %s, in organization: %s",
+		payload.GetRepo().GetName(), payload.GetOrganization().GetLogin())
+
+	if !payload.PullRequest.GetMerged() {
+		wh.logger.Debugf("Ignoring pull request closed event for non-merged pull request #%d, in %s",
+			payload.GetPullRequest().GetNumber(), payload.GetRepo().GetFullName())
+		return
+	}
+
+	pullRequest, err := wh.db.GetPullRequest(&pb.PullRequest{
+		ExternalRepositoryID: uint64(payload.GetRepo().GetID()),
+		Number:               uint64(payload.GetPullRequest().GetNumber()),
+	})
```

Comment:
Check key.

Reply:
This is the third place where I am unsure if just ExternalRepositoryID and Number is sufficient for finding the pr.
Maybe it would be easier to add a ExternalPullRequestID field, to ensure that we get the correct PR.

Reply:
It seems like you can only have one PR per branch, so this should be ok.